pip/pip-379.md

```diff
@@ -0,0 +1,115 @@
+
+### Out of Scope
+
+- Changes to other subscription types (Exclusive, Failover, Shared).
+- Adding support key based ordering guarantees when negative acknowledgements are used
+
+## High-Level Design
+
+### Updated contract of preserving ordering
+
+The "Preserving order of processing" section of the Key_Shared documentation would be updated to contain this contract:
+
+In Key_Shared subscriptions, messages with the same key are delivered and allowed to be in unacknowledged state to only one consumer at a time.
+
+When new consumers join or leave, the consumer handling a message key can change when the default AUTO_SPLIT mode is used, but only after all pending messages for a particular key are acknowledged or the original consumer disconnects.
+
+The Key_Shared subscription doesn't prevent using any methods in the consumer API. For example, the application might call `negativeAcknowledge` or the `redeliverUnacknowledgedMessages` method. When messages are scheduled for delivery due to these methods, they will get redelivered as soon as possible. There's no ordering guarantee in these cases, however the guarantee of delivering a message key to a single consumer at a time will continue to be preserved.
+
+### Future work in needed for supporting key-based ordering with negative acknowledgements
+
+The updated contract explicitly states that it is not possible to retain key-based ordering of messages when negative acknowledgements are used. Changing this is out of scope for PIP-379. A potential future solution for handling this would be to modify the client so that when a message is negatively acknowledged, it would also reject all further messages with the same key until the original message gets redelivered. It's already possible to attempt to implement this in client-side code. However, a proper solution would require support on the broker side to block further delivery of the specific key when there are pending negatively acknowledged messages until all negatively acknowledged messages for that particular key have been acknowledged by the consumer. This solution is out of scope for PIP-379. A future implementation to address these problems could build upon PIP-379 concepts such as "draining hashes" and extend that to cover the negative acknowledgement scenarios.
+
+### High-Level implementation plan
+
+The proposed solution introduces a "draining hashes" concept to efficiently manage message ordering in Key_Shared subscriptions:
+
+1. When consumer hash ranges change (e.g., a consumer joins or leaves), affected hashes of pending messages are added to a "draining hashes" set.
+2. Following messages with hashes in the "draining hashes" set are blocked from further delivery until pending messages are processed.
+3. A reference counter tracks pending messages for each hash in the "draining hashes" set.
+4. As messages are acknowledged or consumers disconnect and therefore get removed from pending messages, the reference counter is decremented.
```

Comment:
Can this method be used when there are some messages with the same **key hash** but different **keys**?
(I think it does because the draining hash data structure holds the number of messages.)

Reply:
The different keys wouldn't matter. This solution is for tracking at hash level. The invariant of "In Key_Shared subscriptions, messages with the same key are delivered and allowed to be in unacknowledged state to only one consumer at a time." would still hold. 
It can be considered an implementation detail that the tracking is at hash level. This would be explained to users since the observability (stats/metrics) is at the hash level.

The broker doesn't unpack batch messages. It's the responsibility of the producer to produce batch messages where all messages in the batch contain the same key and therefore have the same hash.

Reply:
The main consequence of hash collisions is that a single key is coupled to any other key with the same hash and would get blocked together.